packages/grid/_modules_/grid/hooks/features/pagination/useGridPagination.ts

```diff
@@ -66,60 +56,89 @@ export const useGridPagination = (apiRef: GridApiRef): void => {
-        pageSize,
-      });
     },
-    [apiRef, setGridState, forceUpdate, logger, options.page],
+    [setGridState, forceUpdate, logger],
   );
 
-  useGridApiOptionHandler(apiRef, GRID_PAGE_CHANGE, options.onPageChange);
-  useGridApiOptionHandler(apiRef, GRID_PAGE_SIZE_CHANGE, options.onPageSizeChange);
+  React.useEffect(() => {
+    apiRef.current.updateControlState<number>({
+      stateId: 'page',
+      propModel: props.page,
+      propOnChange: props.onPageChange,
+      stateSelector: (state) => state.pagination.page,
+      onChangeCallback: (model) => {
+        apiRef.current.publishEvent(GRID_PAGE_CHANGE, model);
+      },
+    });
+  }, [apiRef, props.page, props.onPageChange]);
+
+  React.useEffect(() => {
+    apiRef.current.updateControlState<number>({
+      stateId: 'pageSize',
+      propModel: props.pageSize,
+      propOnChange: props.onPageSizeChange,
+      stateSelector: (state) => state.pagination.pageSize,
+      onChangeCallback: (model) => {
+        apiRef.current.publishEvent(GRID_PAGE_SIZE_CHANGE, model);
+      },
+    });
+  }, [apiRef, props.pageSize, props.onPageSizeChange]);
+
+  React.useEffect(() => {
+    if (options.page === undefined) {
+      return;
+    }
+
+    setGridState((oldState) => ({
+      ...oldState,
+      pagination: {
+        ...oldState.pagination,
+        page: options.page!,
+      },
+    }));
+  }, [setGridState, options.page]);
+
+  React.useEffect(() => {
+    if (
+      options.pageSize === undefined &&
+      (!options.autoPageSize || !containerSizes?.viewportPageSize)
+    ) {
+      return;
+    }
+
+    setGridState((oldState) => ({
+      ...oldState,
+      pagination: {
+        ...oldState.pagination,
+        pageSize: options.autoPageSize ? containerSizes?.viewportPageSize! : options.pageSize!,
+      },
+    }));
+  }, [setGridState, options.pageSize, containerSizes?.viewportPageSize, options.autoPageSize]);
 
   React.useEffect(() => {
     setGridState((state) => ({
       ...state,
-      pagination: applyConstraints(
-        {
-          ...state.pagination,
-          paginationMode:
-            options.paginationMode != null
-              ? options.paginationMode
-              : state.pagination.paginationMode,
-          rowCount: options.rowCount !== undefined ? options.rowCount : visibleRowCount,
-          pageSize:
-            (options.autoPageSize ? containerSizes?.viewportPageSize : options.pageSize) ||
-            state.pagination.pageSize,
-        },
-        options.page,
-      ),
+      pagination: applyConstraints({
+        ...state.pagination,
+        rowCount: options.rowCount !== undefined ? options.rowCount : visibleRowCount,
+      }),
     }));
     forceUpdate();
   }, [
     setGridState,
     forceUpdate,
-    options.paginationMode,
     visibleRowCount,
     options.rowCount,
     options.autoPageSize,
-    containerSizes?.viewportPageSize,
     options.pageSize,
     options.page,
+    containerSizes?.viewportPageSize,
```

Comment:
why do we need this one now?

Reply:
Because if we updated the state [here](https://github.com/mui-org/material-ui-x/blob/0c733adb212748d20c2c661e99eff1b43fcc566a/packages/grid/_modules_/grid/hooks/features/pagination/useGridPagination.ts#L115) due to a viewport height change.

Then we must check if the pagination state is still valid (= rerun `applyContraints`)
But if we run it directly on the `setGridState` link above. Then we have cases were we update both `pageSize` and `page` in the same `setGridState`. And your hook does not like it.

So the workaround I found is to update `page` and `pageSize` separately.
And then in a third `setGridState` that is run whenever the rwo first are run, I update de constraints.

But it's quite ugly.
If you have a better approach I'm interested.

Reply:
I could remove this specific variable from the 3rd useEffect by checking if we are in `autoPageSize` in the 1st useEffect and apply the constraint directly their if so.
It would be a bit more optimized, but even more hard to read.

Reply:
This is no longer an issue after the split of `usePagination` into two hooks